network/dag/subscriber.go

```diff
@@ -0,0 +1,356 @@
+/*
+ * Copyright (C) 2022 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package dag
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/avast/retry-go/v4"
+	"github.com/nuts-foundation/go-stoabs"
+	"github.com/nuts-foundation/nuts-node/crypto/hash"
+	"github.com/nuts-foundation/nuts-node/network/log"
+	"time"
+)
+
+const (
+	defaultRetryDelay      = time.Second
+	retriesFailedThreshold = 10
+	maxRetries             = 100
+	// TransactionEventType is used as Type in an Event when a transaction is added to the DAG.
+	TransactionEventType = "transaction"
+	// PayloadEventType is used as Type in an Event when a payload is written to the DB.
+	PayloadEventType = "payload"
+)
+
+// Subscriber defines methods for a persistent retry mechanism.
+// Storing the event in the DB is separated from notifying the subscribers.
+// The event is sent to subscribers after the transaction is committed to prevent timing issues.
+type Subscriber interface {
```

Comment:
Given the pub-sub pattern, Subscriber sounds a bit off; this is more like an EventPublisher? (since it does all the event-queue-retry stuff). I'd call only the receiver of the event the subscriber.

Reply:
Shouldn't we actually use NATS for this?

Reply:
> Shouldn't we actually use NATS for this?

We can't use Nats transactionally and Nats then becomes part of "MUST be backed up" family. Saving the extra backup requirement with a few hundred lines of code was worth the investment.

Reply:
Renamed to `(New)Notifier` with a `ReceiverFn`